compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+            """
+            Returns the source of the old source before the for-loop.
+            """
+            while not before_loop(memref_val.owner):
+                memref_val = memref_val.owner.source
+            return memref_val
+
+        def get_dynamic_sizes_and_add(alloc_op, ops_to_add) -> Sequence[SSAValue]:
+            """
+            Returns the dynamic sizes of the alloc operation.
+            If the size is defined before the loop, it is kept as is.
+            If the size can be derived from a constant, the constant is created and added to ops_to_add.
+            """
+            dynamic_sizes = []
+            for size in alloc_op.dynamic_sizes:
+                assert isinstance(size.owner, Operation)
+                if before_loop(size.owner):
+                    dynamic_sizes.append(size)
+                elif isinstance(size.owner, memref.Dim):
+                    new_constant = size.owner.index.owner.clone_without_regions()
+                    ops_to_add.append(new_constant)
+                    source_op = replace_dim(size.owner)
+                    if isinstance(source_op, arith.Constant):
+                        dynamic_sizes.append(source_op.results[0])
+                        ops_to_add.append(source_op)
+                    else:
+                        new_dim = memref.Dim.from_source_and_index(
+                            get_source_before_for_loop(size.owner.source),
+                            new_constant.results[0],
+                        )
+                        ops_to_add.append(new_dim)
+                        dynamic_sizes.append(new_dim.results[0])
+                else:
+                    new_constant = get_constant_value(size.owner)
+                    ops_to_add.append(new_constant)
+                    dynamic_sizes.append(new_constant.results[0])
+            return dynamic_sizes
+
+        # Repeat When Changes are made.
+        changes_made = False
+        # if the alloc can be moved, detach it from the parent and insert new Alloc Object in front of the for-loop
+        # When constants or maximal values where found from inside the loop, we need to insert these before the alloc.
+        # Only the first for-loop is considered, the algorithm can be repeated to elevate the alloc-op higher.
+        if can_move_alloc(alloc_op):
+            changes_made = True
+            ops_to_add = []
+            new_alloc_op = memref.Alloc(
+                dynamic_sizes=get_dynamic_sizes_and_add(alloc_op, ops_to_add),
+                symbol_operands=alloc_op.symbol_operands,
+                result_type=alloc_op.results[0].type,
+                alignment=alloc_op.alignment,
+            )
+            rewriter._replace_all_uses_with(
+                # This is a private function of Rewriter Class and should be replaced with a public function,
+                # but for now it works
+                alloc_op.results[0],
+                new_alloc_op.results[0],
+            )
```

Comment:
Can you not simply use `replace_op` instead?

Reply:
Yes I think all functions starting with `_` are not meant to be used by the outside world